### PR TITLE
Print version information if present in FDT

### PR DESCRIPTION
**Pull request by PeterRugg**
Recent bitfiles have the "sri-cambridge,version" property populated with git commit hashes of the CPUs (in each CPU node) and the SoC (in the root node). This PR prints these on boot so people can tell what bitfile they have.

### CODE REVIEW

File: machine/fdt.c [hunk 1]
```diff
@@ -204,6 +204,64 @@ void query_mem(uintptr_t fdt)
...
+void query_root(uintptr_t fdt)
+{
+  struct fdt_cb cb;
+  struct root_scan scan;
+
+  memset(&cb, 0, sizeof(cb));
+  memset(&scan, 0, sizeof(scan));
+  cb.open = root_open;
+  cb.prop = root_prop;
+  cb.done = NULL;
```
Comment (jrtc27) [inline machine/fdt.c]: Implied by the memset

File: machine/fdt.c [hunk 2]
```diff
@@ -204,6 +204,64 @@ void query_mem(uintptr_t fdt)
...
+  if (!strcmp(prop->name, "sri-cambridge,version") && !scan->version_value) {
+    scan->version_value = prop->value;
+    scan->version_len = prop->len;
+  }
+}
+
+static int root_close(const struct fdt_scan_node *node, void *extra)
+{
+  struct root_scan *scan = (struct root_scan *)extra;
+  if (scan->root && node == scan->root && scan->version_value) {
```
Comment (jrtc27) [inline machine/fdt.c]:
```suggestion
  if (scan->root == node && scan->version_value) {
```
First condition is redundant given node isn't null, second one is an unusual way round

File: machine/fdt.c [hunk 3]
```diff
@@ -255,6 +320,17 @@ static void hart_done(const struct fdt_scan_node *node, void *extra)
 
   if (scan->cpu == node) {
     assert (scan->hart >= 0);
+    if (scan->version_value) {
+      printm("Hart %d version: ", scan->hart);
+      char *char_data = (char *)(scan->version_value);
+      // The size should be 1, but print any extra values if they appear
+      for (size_t i = 0; i < scan->version_len; i += strlen(char_data + i) + 1) {
```
Comment (jrtc27) [inline machine/fdt.c]: Use int, version_len is an int not a size_t

File: machine/fdt.c [hunk 4]
```diff
@@ -255,6 +320,17 @@ static void hart_done(const struct fdt_scan_node *node, void *extra)
 
   if (scan->cpu == node) {
     assert (scan->hart >= 0);
+    if (scan->version_value) {
```
Comment (jrtc27) [inline machine/fdt.c]:
```suggestion

    if (scan->version_value) {
```

File: machine/fdt.c [hunk 5]
```diff
@@ -204,6 +204,64 @@ void query_mem(uintptr_t fdt)
...
+}
+
+static int root_close(const struct fdt_scan_node *node, void *extra)
+{
+  struct root_scan *scan = (struct root_scan *)extra;
+  if (scan->root && node == scan->root && scan->version_value) {
+    printm("SoC version: ");
+    char *char_data = (char *)(scan->version_value);
+    // The size should be 1, but print any extra values if they appear
+    for (size_t i = 0; i < scan->version_len; i += strlen(char_data + i) + 1) {
```
Comment (jrtc27) [inline machine/fdt.c]: As below

File: machine/fdt.c [hunk 6]
```diff
@@ -204,6 +204,64 @@ void query_mem(uintptr_t fdt)
...
+    scan->version_len = prop->len;
+  }
+}
+
+static int root_close(const struct fdt_scan_node *node, void *extra)
+{
+  struct root_scan *scan = (struct root_scan *)extra;
+  if (scan->root && node == scan->root && scan->version_value) {
+    printm("SoC version: ");
+    char *char_data = (char *)(scan->version_value);
```
Comment (jrtc27) [inline machine/fdt.c]: Parens aren't needed

File: machine/fdt.c [hunk 7]
```diff
@@ -255,6 +320,17 @@ static void hart_done(const struct fdt_scan_node *node, void *extra)
 
   if (scan->cpu == node) {
     assert (scan->hart >= 0);
+    if (scan->version_value) {
+      printm("Hart %d version: ", scan->hart);
+      char *char_data = (char *)(scan->version_value);
```
Comment (jrtc27) [inline machine/fdt.c]: As above

File: machine/fdt.c (hunk 2 above)
Comment (jrtc27) [reply machine/fdt.c]: Perhaps we should print "unknown" for these rather than nothing at all if not present?

File: machine/fdt.c [hunk 8]
```diff
@@ -144,6 +144,23 @@ int fdt_string_list_index(const struct fdt_scan_prop *prop, const char *str)
...
+void fdt_version_prop_print(const uint32_t * version_value, int version_len)
+{
+  const char *char_data = (char *)version_value;
+  int char_len = version_len;
+  if (!char_data) {
+    char_data = "unknown";
+    char_len = strlen(char_data) + 1;
+  }
+  // The size should be 1, but print any extra values if they appear
+  for (int i = 0; i < char_len; i += strlen(char_data + i) + 1) {
```
Comment (arichardson) [inline machine/fdt.c]:
```suggestion
  if (!char_data) {
    char_data = "unknown";
    version_len = strlen(char_data) + 1;
  }
  // The size should be 1, but print any extra values if they appear
  for (int i = 0; i < version_len; i += strlen(char_data + i) + 1) {
```

No need for the char_len variable?

File: machine/fdt.c [hunk 9]
```diff
@@ -204,6 +221,56 @@ void query_mem(uintptr_t fdt)
...
+  struct root_scan *scan = (struct root_scan *)extra;
+  if (!scan->root) {
+    scan->root = node;
+  }
+}
+
+static void root_prop(const struct fdt_scan_prop *prop, void *extra)
+{
+  struct root_scan *scan = (struct root_scan *)extra;
+  if (!strcmp(prop->name, "sri-cambridge,version") && !scan->version_value) {
```
Comment (arichardson) [inline machine/fdt.c:242]: I generally find `strcmp(a, b) == 0` to be easier to parse (the ! can be missed quite easily), but don't really mind either way.

Comment (PeterRugg) [reply machine/fdt.c:242]: I definitely agree, but there are 10s of other strcmps in the file done this way, so I figured I'd match the style.

**Review by arichardson — APPROVED** [review]
LGTM but I'd wait for @jrtc27 to have another look.

File: machine/fdt.c [hunk 10]
```diff
@@ -144,6 +144,22 @@ int fdt_string_list_index(const struct fdt_scan_prop *prop, const char *str)
   return -1;
 }
 
+void fdt_version_prop_print(const uint32_t * version_value, int version_len)
```
Comment (jrtc27) [inline machine/fdt.c]:
```suggestion
void fdt_version_prop_print(const uint32_t *version_value, int version_len)
```

File: machine/fdt.c [hunk 11]
```diff
@@ -144,6 +144,22 @@ int fdt_string_list_index(const struct fdt_scan_prop *prop, const char *str)
...
+{
+  const char *char_data = (char *)version_value;
+  if (!char_data) {
+    char_data = "unknown";
+    version_len = strlen(char_data) + 1;
+  }
+  // The size should be 1, but print any extra values if they appear
```
Comment (jrtc27) [inline machine/fdt.c]: This doesn't quite make sense; the size would normally be strlen+1, which is not 1. Maybe "There should only be one string, but print any extra strings if they appear"?

File: machine/fdt.c [hunk 12]
```diff
@@ -144,6 +144,22 @@ int fdt_string_list_index(const struct fdt_scan_prop *prop, const char *str)
   return -1;
 }
 
+void fdt_version_prop_print(const uint32_t * version_value, int version_len)
+{
+  const char *char_data = (char *)version_value;
```
Comment (jrtc27) [inline machine/fdt.c]:
```suggestion
  const char *char_data = (const char *)version_value;
```
since version_value is a `const uint32_t *`

**Review by jrtc27 — APPROVED** [review]
Looks fine, please squash and merge once CI passes

**Pull request merged**